controller/aks-cluster-config-handler.go

```diff
@@ -0,0 +1,728 @@
+	}
+
+	if err = future.WaitForCompletionRef(ctx, groupsClient.Client); err != nil {
+		logrus.Errorf("[AKS] cannot get the AKS cluster create or update future response: %v", err)
+		return err
+	}
+
+	logrus.Infof("Cluster %v removed successfully", spec.ClusterName)
+
+	return nil
+}
+
+func newClustersClient(authorizer autorest.Authorizer, spec aksv1.AKSClusterConfigSpec) (*containerservice.ManagedClustersClient, error) {
+	if authorizer == nil {
+		return nil, fmt.Errorf("error getting Azure authorizer")
+	}
+
+	baseURL := spec.BaseURL
+	if baseURL == "" {
+		baseURL = azure.PublicCloud.ResourceManagerEndpoint
+	}
+
+	client := containerservice.NewManagedClustersClientWithBaseURI(baseURL, spec.SubscriptionID)
+	client.Authorizer = authorizer
+
+	return &client, nil
+}
+
+// CreateAKS creates a new managed Kubernetes cluster
+func createOrUpdateCluster(ctx context.Context, secretsCache wranglerv1.SecretCache, clusterClient *containerservice.ManagedClustersClient, spec aksv1.AKSClusterConfigSpec) (c containerservice.ManagedCluster, err error) {
+
+	dnsPrefix := spec.DNSPrefix
+	if dnsPrefix == "" {
+		dnsPrefix = spec.ClusterName + "-master"
+	}
+
+	tags := make(map[string]*string)
+	for key, val := range spec.Tags {
+		if val != "" {
+			tags[key] = to.StringPtr(val)
+		}
+	}
+	displayName := spec.DisplayName
+	if displayName == "" {
+		displayName = spec.ClusterName
+	}
+	tags["displayName"] = to.StringPtr(displayName)
+
+	kubernetesVersion := spec.KubernetesVersion
+	if kubernetesVersion == "" {
+		kubernetesVersion = "1.18.14"
+	}
+
+	//var vmNetSubnetID *string
+	networkProfile := &containerservice.NetworkProfile{}
+	if hasCustomVirtualNetwork(spec) {
+		//virtualNetworkResourceGroup := spec.ResourceGroup
+
+		// if virtual network resource group is set, use it, otherwise assume it is the same as the cluster
+		//if spec.VirtualNetworkResourceGroup != "" {
+		//	virtualNetworkResourceGroup = spec.VirtualNetworkResourceGroup
+		//}
+		/*
+		vmNetSubnetID = to.StringPtr(fmt.Sprintf(
+			"/subscriptions/%v/resourceGroups/%v/providers/Microsoft.Network/virtualNetworks/%v/subnets/%v",
+			spec.SubscriptionID,
+			virtualNetworkResourceGroup,
+			spec.VirtualNetwork,
+			spec.Subnet,
+		))
+		*/
+		networkProfile.DNSServiceIP = to.StringPtr(spec.NetworkDNSServiceIP)
+		networkProfile.DockerBridgeCidr = to.StringPtr(spec.NetworkDockerBridgeCIDR)
+		networkProfile.ServiceCidr = to.StringPtr(spec.NetworkServiceCIDR)
+
+		if spec.NetworkPlugin == "" {
+			networkProfile.NetworkPlugin = containerservice.Azure
+		} else {
+			networkProfile.NetworkPlugin = containerservice.NetworkPlugin(spec.NetworkPlugin)
+		}
+
+		// if network plugin is 'kubenet', set PodCIDR
+		if networkProfile.NetworkPlugin == containerservice.Kubenet {
+			networkProfile.PodCidr = to.StringPtr(spec.NetworkPodCIDR)
+		}
+
+		if spec.NetworkPolicy != "" {
+			networkProfile.NetworkPolicy = containerservice.NetworkPolicy(spec.NetworkPolicy)
+		}
+	}
+
```

Comment:
Not sure I'm reading it right here but if there is no custom vnet and this is a new cluster the user should still have the ability to pick various vnet options *if* they pick the azure CNI networking option. OTOH, if they pick kubenet azure creates the vnet, subnet etc. for the user.

https://docs.microsoft.com/en-us/azure/aks/configure-azure-cni
 

OTOH from what I have seen this is not how v1 works (it allows users to specify their custom vnet with both kubenet and azure and things still seem to work) so maybe we continue doing that but it _is_ a different ux from azure portal. Maybe something for @cloudnautique or @wjimenez5271 to consider.

Reply:
I have re-added custom vnet option